cmd/clusterctl/client/repository/repository_github.go

```diff
@@ -80,17 +83,43 @@ func injectGithubClient(c *github.Client) githubRepositoryOption {
 	}
 }
 
+func injectGoproxyClient(c *goproxyClient) githubRepositoryOption {
+	return func(g *gitHubRepository) {
+		g.injectGoproxyClient = c
+	}
+}
+
 // DefaultVersion returns defaultVersion field of gitHubRepository struct.
 func (g *gitHubRepository) DefaultVersion() string {
 	return g.defaultVersion
 }
 
 // GetVersions returns the list of versions that are available in a provider repository.
 func (g *gitHubRepository) GetVersions() ([]string, error) {
-	versions, err := g.getVersions()
+	log := logf.Log
+
+	cacheID := fmt.Sprintf("%s/%s", g.owner, g.repository)
```

Comment:
Out of curiosity: How often do we get cache hits?
I can only think of `clusterctl upgrade` that will do more than one call to `GetVersions` (one while creating the github repo client and one while looking for version to upgrade to). Are there any other cases where the cache will be used?

Reply:
* Per provider getting installed we will have **two** calls to GetVersions
* worst case: we reduce API Calls by the caching feature by 50%
* In case of core/cluster-api repo where we install 3 providers (capi, capi-kubeadm-bootstrap, capi-kubeadm-control-plane) we will have only 1 actual request and 5 cache hits + depending on provider 1 actual request and 5 cache hits
* for capd installation (capi, ,capi-kubeadm-bootstrap, capi-kubeadm-control-plane, capd) we will only have 1 actual request and 7 cache hits.

Example for when initializing for aws:

```
❯ AWS_B64ENCODED_CREDENTIALS=$(echo foo | base64) go run ./cmd/clusterctl init --infrastructure aws
Fetching providers
GetVersions UNCACHED: kubernetes-sigs cluster-api
GetVersions CACHEHIT: kubernetes-sigs cluster-api
GetVersions CACHEHIT: kubernetes-sigs cluster-api
GetVersions UNCACHED: kubernetes-sigs cluster-api-provider-aws
GetVersions CACHEHIT: kubernetes-sigs cluster-api
GetVersions CACHEHIT: kubernetes-sigs cluster-api
GetVersions CACHEHIT: kubernetes-sigs cluster-api
GetVersions CACHEHIT: kubernetes-sigs cluster-api-provider-aws
Installing cert-manager Version="v1.9.1"
Waiting for cert-manager to be available...
Installing Provider="cluster-api" Version="v1.2.2" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v1.2.2" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v1.2.2" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-aws" Version="v1.5.0" TargetNamespace="capa-system"
```

Reply:
This is great info. The cluster-api  repo with 3 providers is a really good case. 👍🏼